norse/torch/module/lsnn.py

```diff
@@ -107,25 +107,47 @@ class LSNNLayer(torch.nn.Module):
       cell (torch.nn.Module): the underling neuron module, uninitialized
       *cell_args: variable length input arguments for the underlying cell
                   constructor
+      **cell_kwargs: variable length key-value arguments for the underlying cell constructor
     """
 
-    def __init__(self, cell, *cell_args):
+    def __init__(self, cell, *cell_args, **cell_kwargs):
         super(LSNNLayer, self).__init__()
-        self.cell = cell(*cell_args)
+        self.cell = cell(*cell_args, **cell_kwargs)
 
     def initial_state(self, batch_size, device, dtype=torch.float) -> LSNNState:
         """Return the initial state of the LSNN layer, as given by the
         internal LSNNCell"""
         return self.cell.initial_state(batch_size, device, dtype)
 
     def forward(
-        self, input_tensor: torch.Tensor, state: LSNNState
+        self, input_tensor: torch.Tensor, state: LSNNState, return_states: bool = False
     ) -> Tuple[torch.Tensor, LSNNState]:
+        """
+        Takes one step in the LSNN layer by simulating the layer for a number of timesteps.
+        Since the layer is recurrent, each simulation state (LSNNState) is used as the input for the next step.
+
+        The function expects inputs in the shape (simulation time steps, batch size, ...).
+
+        Parameters:
+        input_tensor (torch.Tensor): Input tensor with timesteps in the first dimension
+        state (LSNNState): The input LSNN state
+        return_states (bool): If True, return a list of all states from all the timesteps. If False, only return the final state after the last timestep.
+
+        Returns:
+        A tuple of all the spikes for each timestep and either
+            a) a single LSNNState, if return_states is set to False (or not set) or
+            b) a list of LSNNStates, if return_states is set to True
+        """
         inputs = input_tensor.unbind(0)
+        states = []
         outputs = []  # torch.jit.annotate(List[torch.Tensor], [])
         for i in range(len(inputs)):
             out, state = self.cell(inputs[i], state)
             outputs += [out]
+            if return_states:
```

Comment:
I think it would be preferable to collect the state of the voltages, currents and b separately and return them as stacked tensors in an LSNNState.

Reply:
I like the idea, but I can see how it's slightly confusing to get a tensor out with an implicit map of v/i/bs. How about reusing the state object then and returning the voltages, currents and biases in the same objects, but stacked? Then the forward pass would always return a `LSNNState` object. We could call the parameter something like `keep_state=False` where each component would be of shape (batches, n) if set to `False`, but shape (timesteps, batches, n) if set to `True`.

Reply:
Alright, this is actually quite annoying because it means we can't use the stacked LSNNState in forward computations. Unless we then later filter out the data. But I don't really see any neat solutions to this. We can either:

* Return a single state / list like before
* Return a single state, but stacked. That will lead to code like this:
 
```python
for ts in ...:
    _, s = lsnn(...)
    s = LSNNState(s.v[-1], s.i[-1], s.b[-1], s.v[-1])
```
* Return a single state with an added triplet if the boolean is set
* ?